dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusVulkan.java

```diff
@@ -0,0 +1,412 @@
+        programBuilder.initVirLocEqLocAliasGen(ctx.location(0).getText(),
+                ctx.location(1).getText());
+        return null;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // SSW declarator list
+    @Override
+    public Object visitSswDeclaratorList(LitmusVulkanParser.SswDeclaratorListContext ctx) {
+        if (ctx != null) {
+            for (LitmusVulkanParser.SswDeclaratorContext sswDeclaratorContext : ctx.sswDeclarator()) {
+                int threadId0 = sswDeclaratorContext.threadId(0).id;
+                int threadId1 = sswDeclaratorContext.threadId(1).id;
+                programBuilder.addSwwPairThreads(threadId0, threadId1);
+            }
+        }
+        return null;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Thread declarator list (on top of instructions)
+    @Override
+    public Object visitThreadDeclaratorList(LitmusVulkanParser.ThreadDeclaratorListContext ctx) {
+        for (LitmusVulkanParser.ThreadScopeContext threadScopeContext : ctx.threadScope()) {
+            int subgroupID = threadScopeContext.subgroupScope().scopeID().id;
+            int workgroupID = threadScopeContext.workgroupScope().scopeID().id;
+            int queuefamilyID = threadScopeContext.queuefamilyScope().scopeID().id;
+            // NB: the order of scopeIDs is important
+            programBuilder.newScopedThread(Arch.VULKAN, threadScopeContext.threadId().id,
+                    queuefamilyID, workgroupID, subgroupID);
+            threadCount++;
+        }
+        return null;
+    }
+
+    // ----------------------------------------------------------------------------------------------------------------
+    // Instruction list (the program itself)
+    @Override
+    public Object visitConstant(LitmusVulkanParser.ConstantContext ctx) {
+        return ExpressionFactory.getInstance().parseValue(ctx.getText(), archType);
+    }
+
+    @Override
+    public Object visitRegister(LitmusVulkanParser.RegisterContext ctx) {
+        return programBuilder.getOrNewRegister(mainThread, ctx.getText(), archType);
+    }
+
+    @Override
+    public Object visitInstructionRow(LitmusVulkanParser.InstructionRowContext ctx) {
+        for (int i = 0; i < threadCount; i++) {
+            mainThread = i;
+            visitInstruction(ctx.instruction(i));
+        }
+        return null;
+    }
+
+    @Override
+    public Object visitStoreConstant(LitmusVulkanParser.StoreConstantContext ctx) {
+        MemoryObject object = programBuilder.getOrNewMemoryObject(ctx.location().getText());
+        IConst constant = (IConst) ctx.constant().accept(this);
+        Boolean atomic = ctx.atomic().isAtomic;
+        String mo = ctx.mo().content;
+        String avvis = ctx.avvis().content;
+        String scope = ctx.scope().content;
```

Comment:
Updated

Reply:
I still don't link how this look. Many times you simply compute `""` as a tag and add this to the event in `tagControl` which brings now value. 

I think it would be cleaner to do the following:
```
Store store;
if(ctx.mo() != null) {
  store = EventFactory.newStoreWithMo(object, value, mo);
} else {
  store = EventFactory.newStore(object, value);
}
if(ctx.avvis() != null) {
  store.addTag(ctx.avvis().content)
... same for scope
```
Make `tagControl` add the remaining tags

Reply:
We should definitely avoid adding the empty string as a tag. If it makes thinks easier, the `addTag` method could be made to automatically ignore empty strings. But I think Hernán's solution is the cleanest.